contents/go/Go.md

```diff
@@ -0,0 +1,143 @@
+
+### Arrays and slices
+Arrays are not often seen in Go code. This is because the size of a Go array must be declared at its creation, which limits how flexible an array can be.
+
+To create dynamically-sized arrays, Go introduces the concept of a slice. Slices are not arrays; rather, they are data structures that describe a piece of a separately-stored array. Slices are the topic of several Go blog entries; you can read more about [the internal layout of slices](https://blog.golang.org/go-slices-usage-and-internals) and [the mechanics of using slices](https://blog.golang.org/slices). Exercises on how to use slices are available in [A Tour of Go](https://tour.golang.org/moretypes/7).
+
+### `defer`, `panic` and `recover`
+In addition to traditional control flow mechanisms such as `if`, `for` and `switch`, Go also provides three other ways to alter the flow of a program: `defer`, `panic`, `recover`. These mechanisms can be conceived in terms of exception handling in Java and C++, but also have more general purpose uses.
+
+- `defer` pushes a function call to a list, and executes all functions on the list in last-in-first-out order after the surrounding function returns. It is frequently used for clean-up actions, such as to close files.
+- `panic` stops the ordinary flow of execution, executes any deferred functions, and returns to its caller as a call to `panic`. The `panic` continues up the stack until all functions in the goroutine have returned, after which the program crashes.
+- `recover` regains control of a panicking goroutine. When used inside a deferred function, a call to recover captures the value returned by `panic` and resumes normal execution.
+
+More information on these three mechanisms and their uses can be found on the [Go blog](https://blog.golang.org/defer-panic-and-recover).
+
+### Interfaces
+Although Go has types and methods and allows pseudo-object-oriented style of programming, type hierarchy does not exist in Go. Instead, Go uses interfaces to specify methods that types should implement, grouping types by functionality rather than using notions of inheritance. Types implement interfaces by implementing the methods in the interface, and do not need to explicitly specify which interfaces are implemented.
+
+In the example below, the `Rectangle` type implements the interface `TwoDimensional` by implementing the methods `area()` and `perim()` that are specified in the interface.
+
+```go
+type TwoDimensional interface {
+  area() float64
+  perim() float64
+}
+
+type Rectangle struct {
+  width, height float64
+}
+
+func (r Rectangle) area() {
+  return r.width * r.height
+}
+
+func (r Rectangle) perim() {
+  return r.width * 2 + r.height * 2
+}
+```
```

Comment:
Add an example, like `type Circle struct` which implements `area()` but not `perim()` to show a negative example.